[ntcore] Create time synchronization protocol (#7007)
wpilibsuite/allwpilib: ntcore/doc/time-sync-protocol.adoc

Clients shall periodically (e.g. every few seconds) send, in a manner that minimizes transmission delays, a **TSP Ping Message** that contains the client's current local time.

When the server recieves a **TSP Ping Message** from any client, it shall respond to the client, in a manner that minimizes transmission delays, with a **TSP Pong message** encoding a timestamp of its (the server's) current local time (in microseconds), and the client-provided data value.

When the client receives a **TSP Pong Message** from the server, it shall compute the round trip time (RTT) from the delta between the message's data value and the current local time.  If the RTT is less than that from previous measurements, the client shall use the timestamp in the message plus ½ the RTT as the server time equivalent to the current local time, and use this equivalence to compute server time base timestamps from local time for future messages.

[[transport]]
=== Transport

Communication between server and clients shall occur over the User Datagram Protocol (UDP) Port 5810.

[[format]]
=== Message Format

> TODO: Should I have an arbitrary magic value? do i need checksums? should we include a sequence number or other way to check if a message was replied to besides checking if we got the timestamp we expected back?

Reviewer: The client can treat the timestamp as a sequence number

Author: Implemented using this, and having the client throw out all pongs that don't match the currently active ping. This does mean that if you send pings faster than the server can handle, you might just never get your reply back since it'll get swapped out, which isn't great.

= Time Syncronization Protocol Specification, Version 1.0
WPILib Developers <wpilib@wpi.edu>
Protocol Revision 1.0, 08/25/2024
:toc:
:toc-placement: preamble
:sectanchors:

[[roles]]
=== Roles

> Roles assumed by Time Syncronization Protocol (TSP) participants consists of exactly one server and zero or more clients per network. The server role is responsible for listening for incoming time synchronization requests from clients and replying approriately. The client role is responsible for sending "Ping" messages to the server and listening for "Pong" replies to estimate the offset between the server and client time bases. 

Reviewer 2: ```suggestion
Time Syncronization Protocol (TSP) participants can assume either a server role or a client role. Each network consists of exactly one server and zero or more clients. The server role is responsible for listening for incoming time synchronization requests from clients and replying approriately. The client role is responsible for sending "Ping" messages to the server and listening for "Pong" replies to estimate the offset between the server and client time bases.
```
Might just be personal preference, but it seems clearer to separate the statements about the roles of partipants and about the number of each role in a network.

Author: We should probably also define what "network" means

> All time values shall use units of micrseconds. The epoch of the time base this is measured against is unspecified.

Reviewer 3: ```suggestion
All time values shall use units of microseconds. The epoch of the time base this is measured against is unspecified.
```
